README.md
# Star_Print_Gendroid

Gendroid Impresoras Star Micronics
GEN
Seleccionado nota original y copia de la nota y reimpresion.

Emulation
SM-T300i 2.4 or later Bluetooth EscPosMobile

- Agregar texto;

```
    String _texto = "        Star Clothing Boutique\n" +
      "             123 Star Road\n" +
      "           City, State 12345\n" +
      "\n" +
      "Date:MM/DD/YYYY          Time:HH:MM PM\n" +
      "--------------------------------------\n" +
      "SALE\n" +
      "SKU            Description       Total\n" +
      "300678566      PLAIN T-SHIRT     10.99\n" +
      "300692003      BLACK DENIM       29.99\n" +
      "300651148      BLUE DENIM        29.99\n" +
      "300642980      STRIPED DRESS     49.99\n" +
      "30063847       BLACK BOOTS       35.99\n" +
      "\n" +
      "Subtotal                        156.95\n" +
      "Tax                               0.00\n" +
      "--------------------------------------\n" +
      "Total                           156.95\n" +
      "--------------------------------------\n" +
      "\n" +
      "Charge\n" +
      "156.95\n" +
      "Visa XXXX-XXXX-XXXX-0123\n" +
      "Refunds and Exchanges\n" +
      "Within 30 days with receipt\n" +
      "And tags attached\n";

  PrintCommands commands = PrintCommands();
  commands.appendBitmapText(text: texto);
  await Impresora.imprimir(commands: commands);

```

- Agregar UnitList

```
  PrintCommands commands = PrintCommands();
  commands.appendBitmapByte(byteData: )
  await Impresora.imprimir(commands: commands);
```

- Agregar Permission manifest de tu proyecto

```
    <uses-permission android:name="android.permission.BLUETOOTH" />
    <uses-permission android:name="android.permission.BLUETOOTH_ADMIN" />
    <uses-permission android:name="android.permission.BLUETOOTH_CONNECT" />
    <uses-permission android:name="android.permission.ACCESS_FINE_LOCATION" />
    <uses-permission android:name="android.permission.BLUETOOTH_SCAN" />
```

- Agregar PermissionHandler

```
  final permison = await Permission.location.request();
  final blue = await Permission.bluetoothConnect.request();
  final blueS = await Permission.bluetoothScan.request();
```


CHANGELOG.md
## 0.0.1

- TODO: Describe initial release.

## 0.0.2

- TODO: Describe initial release.

## 0.0.3

- TODO: Con copia y reimpresion de nota con distintivos logos

## 0.0.4

- TODO: Imprime nota con opciones de reimpresion de nota

## 0.0.5

- TODO: Imprime nota con opciones asistencia.

## 0.0.6

- TODO: Solucion espacio al cortar la nota.

## 0.0.7

- TODO: Cambia el Gradle Android.

## 0.0.8

- TODO: Cambia el Gradle Android 7.1.

## 0.0.9

- TODO: Cambia el Gradle Android 7.1. Java 11

## 0.1.0

- TODO: Cambia el Gradle Android 7.1. Java 11

## 0.2.0

- TODO: Update Dart 3 to android 12


lib/src/star_prnt.dart
import 'dart:async';
import 'package:flutter/services.dart';
import 'package:star_print_gen/star_print_gen.dart';

/// Class to handle printer communication
class StarPrnt {
  static const MethodChannel _channel =
      const MethodChannel('app.condominios/impresora');

  /// Scan for available printers. Have specify [StarPortType] of the printer
  static Future<List<PortInfo>> portDiscovery(StarPortType portType) async {
    dynamic result =
        await _channel.invokeMethod('portDiscovery', {'type': portType.text});
    if (result is List) {
      return result.map<PortInfo>((port) {
        return PortInfo(port);
      }).toList();
    } else {
      return [];
    }
  }

  /// Check status of printer. Have specify [portName] and [emulation]. Returns [PrinterResponseStatus]. Use [StarMicronicsUtilities] to find suitable emulations.
  static Future<PrinterResponseStatus> getStatus({
    required String portName,
    required String emulation,
  }) async {
    dynamic result = await _channel.invokeMethod('checkStatus', {
      'portName': portName,
      'emulation': emulation,
    });
    return PrinterResponseStatus.fromMap(
      Map<String, dynamic>.from(result),
    );
  }

  /// Sends [PrintCommands] to the printer. Have to specify [portName] and [emulation]. Returns [PrinterResponseStatus]
  static Future<PrinterResponseStatus> sendCommands({
    
    required String portName,
    required String emulation,
    required PrintCommands printCommands,
    bool logo = true,
    bool asistencia = false,
    bool reimpresion = false,
    bool copia = false,
  }) async {
    dynamic result = await _channel.invokeMethod('print', {
      'portName': portName,
      'emulation': emulation,
      'printCommands': printCommands.getCommands(),
      'reimpresion': reimpresion,
      'copia': copia,
      'logo' : logo,
      'asistencia': asistencia
    });
    return PrinterResponseStatus.fromMap(
      Map<String, dynamic>.from(result),
    );
  }

  /// sends commands to printer to run
  @Deprecated('Use sendCommands instead.')
  static Future<dynamic> print({
    required String portName,
    required String emulation,
    required PrintCommands printCommands,
  }) async {
    dynamic result = await _channel.invokeMethod('print', {
      'portName': portName,
      'emulation': emulation,
      'printCommands': printCommands.getCommands(),
    });
    return result;
  }

  /// Check status of printer
  @Deprecated('Use getStatus instead.')
  static Future<dynamic> checkStatus({
    required String portName,
    required String emulation,
  }) async {
    dynamic result = await _channel.invokeMethod('checkStatus', {
      'portName': portName,
      'emulation': emulation,
    });
    return result;
  }

  // static Future<dynamic> connect({
  //   required String portName,
  //   required String emulation,
  //   bool hasBarcodeReader = false,
  // }) async {
  //   dynamic result = await _channel.invokeMethod('connect', {
  //     'portName': portName,
  //     'emulation': emulation,
  //     'hasBarcodeReader': hasBarcodeReader,
  //   });
  //   return result;
  // }
}


lib/src/print_commands.dart
import 'dart:typed_data';
import 'dart:ui' as ui;

import 'package:flutter/rendering.dart';
import 'package:flutter/widgets.dart';
import 'enums.dart';

class PrintCommands {
  /// list of commands to pass to the printer
  List<Map<String, dynamic>> _commands = [];

  /// get current list of commands
  List<Map<String, dynamic>> getCommands() {
    return _commands;
  }

  /// sets the encoding for text prints, [StarEncoding] available is different for each printer
  appendEncoding(StarEncoding encoding) {
    this._commands.add({"appendEncoding": encoding.text});
  }

  /// run command on printer cuter, [StarCutPaperAction] available is different for each printer
  appendCutPaper(StarCutPaperAction action) {
    this._commands.add({"appendCutPaper": action.text});
  }

  /// open cash drawer, [actionNumber] needed to based on the printer port
  openCashDrawer(int actionNumber) {
    this._commands.add({"openCashDrawer": actionNumber});
  }

  /// Prints an image with a url or a file [path].
  /// Set [bothScale] to scale the image by the [width] of receipt.
  /// Sets [absolutePosition] image absolute position.
  /// [StarAlignmentPosition] sets image alignment.
  /// [StarBitmapConverterRotation] set image rotation.
  appendBitmap({
    required String path,
    bool diffusion = true,
    int width = 576,
    bool bothScale = true,
    int? absolutePosition,
    StarAlignmentPosition? alignment,
    StarBitmapConverterRotation? rotation,
  }) {
    Map<String, dynamic> command = {
      "appendBitmap": path,
    };
    command['bothScale'] = bothScale;
    command['diffusion'] = diffusion;
    command['width'] = width;
    if (absolutePosition != null)
      command['absolutePosition'] = absolutePosition;
    if (alignment != null) command['alignment'] = alignment.text;
    if (rotation != null) command['rotation'] = rotation.text;

    this._commands.add(command);
  }

  /// Prints an image with a raw data [byteData].
  /// Set [bothScale] to scale the image by the [width] of receipt.
  /// Sets [absolutePosition] image absolute position.
  /// [StarAlignmentPosition] sets image alignment.
  /// [StarBitmapConverterRotation] set image rotation.
  appendBitmapByte({
    required Uint8List byteData,
    bool diffusion = true,
    int width = 576,
    bool bothScale = true,
    int? absolutePosition,
    StarAlignmentPosition? alignment,
    StarBitmapConverterRotation? rotation,
  }) {
    Map<String, dynamic> command = {
      "appendBitmapByteArray": byteData,
    };
    command['bothScale'] = bothScale;
    command['diffusion'] = diffusion;
    command['width'] = width;
    if (absolutePosition != null)
      command['absolutePosition'] = absolutePosition;
    if (alignment != null) command['alignment'] = alignment.text;
    if (rotation != null) command['rotation'] = rotation.text;

    this._commands.add(command);
  }

  /// Prints an image generated by widgets [widget].
  /// Set [bothScale] to scale the image by the [width] of receipt.
  /// Sets [absolutePosition] image absolute position.
  /// [StarAlignmentPosition] sets image alignment.
  /// [StarBitmapConverterRotation] set image rotation.
  /// Set [Duration] if you are using async widgets or widget that take time to fully build.
  /// logicalSize [Size] is the size of the device the widget is made into.
  /// imageSize [Size] is the size of image generated.
  /// sets the [TextDirection].
  appendBitmapWidget({
    required BuildContext context,
    required Widget widget,
    bool diffusion = true,
    int width = 576,
    bool bothScale = true,
    int? absolutePosition,
    StarAlignmentPosition? alignment,
    StarBitmapConverterRotation? rotation,
    Duration? wait,
    Size? logicalSize,
    Size? imageSize,
    TextDirection textDirection = TextDirection.ltr,
  }) {
    createImageFromWidget(
      context,
      widget,
      wait: wait,
      logicalSize: logicalSize,
      imageSize: imageSize,
      textDirection: textDirection,
    ).then((byte) {
      if (byte != null) {
        appendBitmapByte(
          byteData: byte,
          diffusion: diffusion,
          width: width,
          bothScale: bothScale,
          absolutePosition: absolutePosition,
          alignment: alignment,
          rotation: rotation,
        );
      } else {
        throw Exception('Error generating image');
      }
    });
  }

  /// Prints an image generated by text [text].
  /// [fontSize] sets font text size of image
  /// Set [bothScale] to scale the image by the [width] of receipt.
  /// Sets [absolutePosition] image absolute position.
  /// [StarAlignmentPosition] sets image alignment.
  /// [StarBitmapConverterRotation] set image rotation.
  appendBitmapText({
    required String text,
    int? fontSize,
    bool diffusion = true,
    int? width,
    bool bothScale = true,
    int? absolutePosition,
    StarAlignmentPosition? alignment,
    StarBitmapConverterRotation? rotation,
  }) {
    Map<String, dynamic> command = {
      "appendBitmapText": text,
    };
    command['bothScale'] = bothScale;
    command['diffusion'] = diffusion;
    if (fontSize != null) command['fontSize'] = fontSize;
    if (width != null) command['width'] = width;
    if (absolutePosition != null)
      command['absolutePosition'] = absolutePosition;
    if (alignment != null) command['alignment'] = alignment.text;
    if (rotation != null) command['rotation'] = rotation.text;

    this._commands.add(command);
  }

  /// pushes a manual [command] into the command list
  push(Map<String, dynamic> command) {
    this._commands.add(command);
  }

  /// clear all commands in command list
  clear() {
    this._commands.clear();
  }

  /// Generats an image from [widget].
  /// Set [Duration] if you are using async widgets or widget that take time to fully build.
  /// logicalSize [Size] is the size of the device the widget is made into.
  /// imageSize [Size] is the size of image generated.
  /// sets the [TextDirection].
  static Future<Uint8List?> createImageFromWidget(
    BuildContext context,
    Widget widget, {
    Duration? wait,
    Size? logicalSize,
    Size? imageSize,
    TextDirection textDirection = TextDirection.ltr,
  }) async {
    final RenderRepaintBoundary repaintBoundary = RenderRepaintBoundary();

    logicalSize ??=
        View.of(context).physicalSize / View.of(context).devicePixelRatio;
    imageSize ??= View.of(context).physicalSize;
    assert(logicalSize.aspectRatio == imageSize.aspectRatio);
    final RenderView renderView = RenderView(
      view: WidgetsFlutterBinding.ensureInitialized()
          .platformDispatcher
          .views
          .first,
      child: RenderPositionedBox(
        alignment: Alignment.center,
        child: repaintBoundary,
      ),
      configuration: ViewConfiguration(
        size: logicalSize,
        devicePixelRatio: 1.0,
      ),
    );

    final PipelineOwner pipelineOwner = PipelineOwner();
    final BuildOwner buildOwner = BuildOwner();

    pipelineOwner.rootNode = renderView;
    renderView.prepareInitialFrame();

    final RenderObjectToWidgetElement<RenderBox> rootElement =
        RenderObjectToWidgetAdapter<RenderBox>(
      container: repaintBoundary,
      child: Directionality(
        textDirection: textDirection,
        child: IntrinsicHeight(child: IntrinsicWidth(child: widget)),
      ),
    ).attachToRenderTree(buildOwner);

    buildOwner.buildScope(rootElement);

    if (wait != null) {
      await Future.delayed(wait);
    }

    buildOwner
      ..buildScope(rootElement)
      ..finalizeTree();

    pipelineOwner
      ..flushLayout()
      ..flushCompositingBits()
      ..flushPaint();

    final ui.Image image = await repaintBoundary.toImage(
      pixelRatio: imageSize.width / logicalSize.width,
    );
    final ByteData? byteData =
        await image.toByteData(format: ui.ImageByteFormat.png);

    return byteData?.buffer.asUint8List();
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_star_prnt_example/impresora.dart';
import 'package:permission_handler/permission_handler.dart';
import 'package:star_print_gen/star_print_gen.dart';

void main() async {
  WidgetsFlutterBinding.ensureInitialized();
  final permison = await Permission.location.request();
  final blue = await Permission.bluetoothConnect.request();
  final blueS = await Permission.bluetoothScan.request();

  print("permison ${blue.isGranted}");
  print("blue ${permison.isGranted}");
  print("blueS ${blue.isGranted}");

  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  bool isLoading = false;
  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(title: const Text('Plugin example app')),
        body: Center(
          child: Column(
            mainAxisSize: MainAxisSize.min,
            children: <Widget>[
              TextButton(
                onPressed: () async {
                  PrintCommands commands = PrintCommands();

                  commands.appendBitmapText(text: texto);
                  await Impresora.imprimir(commands: commands);
                  await Impresora.imprimir(
                    reimpresion: true,
                    commands: commands,
                  );
                },
                child: Text('Print from text'),
              ),
            ],
          ),
        ),
      ),
    );
  }
}

const String texto = "        Star Clothing Boutique\n" +
    "             123 Star Road\n" +
    "           City, State 12345\n" +
    "\n" +
    "Date:MM/DD/YYYY          Time:HH:MM PM\n" +
    "--------------------------------------\n" +
    "SALE\n" +
    "SKU            Description       Total\n" +
    "300678566      PLAIN T-SHIRT     10.99\n" +
    "300692003      BLACK DENIM       29.99\n" +
    "300651148      BLUE DENIM        29.99\n" +
    "300642980      STRIPED DRESS     49.99\n" +
    "30063847       BLACK BOOTS       35.99\n" +
    "\n" +
    "Subtotal                        156.95\n" +
    "Tax                               0.00\n" +
    "--------------------------------------\n" +
    "Total                           156.95\n" +
    "--------------------------------------\n" +
    "\n" +
    "Charge\n" +
    "156.95\n" +
    "Visa XXXX-XXXX-XXXX-0123\n" +
    "Refunds and Exchanges\n" +
    "Within 30 days with receipt\n" +
    "And tags attached\n";
